File: SESSION_31/Practice.py
class DNode:
    def __init__(self, initData:int):
        self.data = initData
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.headNode = DNode(None)

    def insertStart(self, newData:int):
        newNode = DNode(newData)
        newNode.next = self.headNode.next
        newNode.prev = self.headNode
        if self.headNode.next != None and type(self.headNode.next) == DNode:
            self.headNode.next.prev = newNode
        self.headNode.next = newNode

    def insertEnd(self, newData:int):
        run = self.headNode
        while run.next != None:
            run = run.next
        run.next = DNode(newData)
        run.next.prev = run

    def showList(self):
        run = self.headNode.next
        print("[START]<->", end='')
        while run != None:
            print(f'[{run.data}]<->', end='')
            run = run.next
        print('[END]')

    def searchNode(self, existingData:int):
        run = self.headNode.next
        while run != None:
            if run.data == existingData:
                return run
            run = run.next
        return None

    def insertAfter(self, existingData:int, newData:int):
        existingNode = self.searchNode(existingData)
        if existingNode == None:
            raise ValueError(f'Existing Data not found {existingData}')
        newNode = DNode(newData)
        newNode.next = existingNode.next
        newNode.prev = existingNode
        if existingNode.next != None:
            existingNode.next.prev = newNode
        existingNode.next = newNode

    def insertBefore(self, existingData:int, newData:int):
        existingNode = self.searchNode(existingData)
        if existingNode == None:
            raise ValueError(f'Existing Data not found {existingData}')
        newNode = DNode(newData)
        newNode.next = existingNode
        newNode.prev = existingNode.prev
        existingNode.prev.next = newNode
        existingNode.prev = newNode

L = DoublyLinkedList()

for data in range(1,6):
    L.insertStart(10*data)

for data in range(6,11):
    L.insertEnd(10*data)

L.insertAfter(100,120)
L.insertBefore(50, 1000)

L.showList()

File: CPPDSA/09_DoublyCircularLinkedList/Practice.cpp
#include <cstdlib>
#include <iostream>

class DNode {
public:
  int data;
  DNode *next;
  DNode *prev;

  DNode(int newData) : data(newData) {
    next = nullptr;
    prev = nullptr;
  }

  DNode() {};
};

class DCL {
private:
  DNode *headNode;

public:
  DCL() {
    headNode = new DNode();
    headNode->next = headNode;
    headNode->prev = headNode;

    std::cout << "headNode is created : " << headNode << std::endl;
  }

  ~DCL() {
    DNode *run = headNode->next;

    while (run != headNode) {
      DNode *runNext = run->next;
      std::cout << "We are freeing the node : " << run << std::endl;
      delete run;
      run = runNext;
    }

    std::cout << "We are deleting the headNode : " << run << std::endl;
    delete run;
  }

  void insertStart(int num);
  void insertEnd(int num);
  int pop();
  void printList();
};

void DCL::insertStart(int num) {
  DNode *newNode = new DNode(num);

  newNode->next = headNode->next;
  newNode->prev = headNode;

  headNode->next = newNode;
  newNode->next->prev = newNode;
}

void DCL::insertEnd(int num) {
  DNode *newNode = new DNode(num);

  newNode->next = headNode;
  newNode->prev = headNode->prev;

  headNode->prev->next = newNode;
  headNode->prev = newNode;
}

int DCL::pop() {
  DNode *run = headNode->next;

  if (run == headNode) {
    std::cout << "Stack is empty" << std::endl;
    exit(EXIT_FAILURE);
  }

  int popNum = run->data;
  headNode->next = run->next;
  run->next->prev = headNode;

  delete run;

  return (popNum);
}

void DCL::printList() {
  DNode *run = headNode->next;

  std::cout << "[START]-->";
  while (run != headNode) {
    std::cout << "[" << run->data << "]" << "-->";
    run = run->next;
  }
  std::cout << "[END]" << std::endl;
}

int main(void) {
  DCL dl;

  for (int i = 1; i < 11; i++) {
    dl.insertStart(i);
  }

  for (int i = 2; i < 21; i++) {
    dl.insertEnd(i);
  }

  dl.printList();

  for (int i = 0; i < 10; i++) {
    int a = dl.pop();
    std::cout << "Popped Element is : " << a << std::endl;
  }

  dl.printList();

  return (0);
}
